src/libstore/signature/signer.cc

```diff
@@ -0,0 +1,85 @@
+#include "signature/signer.hh"
+#include "util.hh"
+
+#include <curl/curl.h>
+#include <sodium.h>
+
+namespace nix {
+    // Splits a signature, that is `$publicKey:$signatureDigest` into a pair (publicKey, signatureDigest)
+    static std::pair<std::string_view, std::string_view> split(std::string_view s)
+    {
+        size_t colon = s.find(':');
+        if (colon == std::string::npos || colon == 0)
+            return {"", ""};
+        return {s.substr(0, colon), s.substr(colon + 1)};
+    }
+
+    Signer::Signer() : pubkey(this->getPublicKey()) { }
+    Signer::Signer(PublicKey && pubkey) : pubkey(pubkey) { }
+
+    bool Signer::verifyDetached(const std::string & data, const std::string & sig) {
+        auto ss = split(sig);
+
+        if (std::string(ss.first) != pubkey.key) return false;
+
+        auto sig2 = base64Decode(ss.second);
+        if (sig2.size() != crypto_sign_BYTES)
+            throw Error("signature is not valid");
+
+        return crypto_sign_verify_detached((unsigned char *) sig2.data(),
+            (unsigned char *) data.data(), data.size(),
+            (unsigned char *) pubkey.key.data()) == 0;
+    }
+
+    const PublicKey& Signer::getPublicKey() const {
+        return pubkey;
+    }
+
+    LocalSigner::LocalSigner(SecretKey && privkey) : Signer(privkey.toPublicKey()), privkey(privkey) {
+    }
+
+    std::string LocalSigner::signDetached(std::string_view s) const {
+        return privkey.signDetached(s);
+    }
+
+    RemoteSigner::RemoteSigner(const std::string & remoteServerUrl) : Signer(),
+        serverUrl(remoteServerUrl),
+        _curl_handle(std::unique_ptr<CURL, decltype(&curl_easy_cleanup)>(curl_easy_init(), curl_easy_cleanup)) {
+        if (!_curl_handle) {
+            throw Error("Failed to initialize curl");
+        }
+
+        curl_easy_setopt(_curl_handle.get(), CURLOPT_USERAGENT, "Nix $version");
```

Comment:
We should probably be doing the same as libstore here:
https://github.com/NixOS/nix/blob/78e886bc5fd9e4d85f8503799540c0b71bb270be/src/libstore/filetransfer.cc#L297-L299

Reply:
(and sharing the same code)

Reply:
Maybe we should even use FileTransfer here.

Reply:
I wanted to avoid using FileTransfer because it seems very tailored to downloading/uploading and this is not what we are doing, the thing is quite massive, and I fear that wiring this up would make things quite hard for me.

Reply:
I proposed a change here, if you want me to use FileTransfer, I would probably need some sync time to discuss how we want to proceed because there would be some stuff to do here.